File: objectModel/Python/cdm/objectmodel/projections/cdm_operation_add_supporting_attribute.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

from typing import Optional, TYPE_CHECKING

from cdm.enums import CdmAttributeContextType, CdmObjectType, CdmOperationType, CdmLogCode
from cdm.objectmodel import CdmAttributeContext
from cdm.resolvedmodel.projections.projection_attribute_state import ProjectionAttributeState
from cdm.utilities import logger, AttributeContextParameters

from .cdm_operation_base import CdmOperationBase

if TYPE_CHECKING:
    from cdm.objectmodel import CdmCorpusContext, CdmTypeAttributeDefinition, CdmTraitReferenceBase
    from cdm.resolvedmodel.projections.projection_attribute_state_set import ProjectionAttributeStateSet
    from cdm.resolvedmodel.projections.projection_context import ProjectionContext
    from cdm.utilities import VisitCallback, ResolveOptions


class CdmOperationAddSupportingAttribute(CdmOperationBase):
    """Class to handle AddSupportingAttribute operations"""

    def __init__(self, ctx: 'CdmCorpusContext') -> None:
        super().__init__(ctx)

        self._TAG = CdmOperationAddSupportingAttribute.__name__
        self.supporting_attribute = None  # type: Optional[CdmTypeAttributeDefinition]
        self.type = CdmOperationType.ADD_SUPPORTING_ATTRIBUTE  # type: CdmOperationType

    def copy(self, res_opt: Optional['ResolveOptions'] = None, host: Optional['CdmOperationAddSupportingAttribute'] = None) -> 'CdmOperationAddSupportingAttribute':
        if not res_opt:
            res_opt = ResolveOptions(wrt_doc=self, directives=self.ctx.corpus.default_resolution_directives)

        copy = CdmOperationAddSupportingAttribute(self.ctx) if not host else host

        copy.supporting_attribute = self.supporting_attribute.copy(res_opt) if self.supporting_attribute else None

        self._copy_proj(res_opt, copy)
        return copy

    def get_name(self) -> str:
        return 'operationAddSupportingAttribute'

    @property
    def object_type(self) -> 'CdmObjectType':
        return CdmObjectType.OPERATION_ADD_SUPPORTING_ATTRIBUTE_DEF

    def validate(self) -> bool:
        missing_fields = []

        if not bool(self.supporting_attribute):
            missing_fields.append('supporting_attribute')

        if len(missing_fields) > 0:
            logger.error(self.ctx, self._TAG, 'validate', self.at_corpus_path, CdmLogCode.ERR_VALDN_INTEGRITY_CHECK_FAILURE, self.at_corpus_path, ', '.join(map(lambda s: '\'' + s + '\'', missing_fields)))
            return False
        return True

    def visit(self, path_from: str, pre_children: 'VisitCallback', post_children: 'VisitCallback') -> bool:
        path = self._fetch_declared_path(path_from)

        if pre_children and pre_children(self, path):
            return False

        if self.supporting_attribute and self.supporting_attribute.visit('{}/supportingAttribute/'.format(path), pre_children, post_children):
            return True

        if post_children and post_children(self, path):
            return True

        return False

    def _append_projection_attribute_state(self, proj_ctx: 'ProjectionContext', proj_output_set: 'ProjectionAttributeStateSet', attr_ctx: 'CdmAttributeContext') -> 'ProjectionAttributeStateSet':
        # Pass through all the input projection attribute states if there are any
        for current_PAS in proj_ctx._current_attribute_state_set._states:
            proj_output_set._add(current_PAS)

        # Create a new attribute context for the operation
        attr_ctx_op_add_supporting_attr_param = AttributeContextParameters()
        attr_ctx_op_add_supporting_attr_param._under = attr_ctx
        attr_ctx_op_add_supporting_attr_param._type = CdmAttributeContextType.OPERATION_ADD_SUPPORTING_ATTRIBUTE
        attr_ctx_op_add_supporting_attr_param._name = 'operation/index{}/{}'.format(self._index, self.get_name())
        attr_ctx_op_add_supporting_attr = CdmAttributeContext._create_child_under(proj_ctx._projection_directive._res_opt, attr_ctx_op_add_supporting_attr_param)

        # Create a new attribute context for the supporting attribute we will create
        attr_ctx_supporting_attr_param = AttributeContextParameters()
        attr_ctx_supporting_attr_param._under = attr_ctx_op_add_supporting_attr
        attr_ctx_supporting_attr_param._type = CdmAttributeContextType.ADDED_ATTRIBUTE_SUPPORTING
        attr_ctx_supporting_attr_param._name = self.supporting_attribute.name
        attr_ctx_supporting_attr = CdmAttributeContext._create_child_under(proj_ctx._projection_directive._res_opt, attr_ctx_supporting_attr_param)
        
        # TODO: this if statement keeps the functionality the same way it works currently in resolution guidance.
        # This should be changed to point to the foreign key attribute instead. 
        # There has to be some design decisions about how this will work and will be done in the next release.
        if len(proj_ctx._current_attribute_state_set._states) > 0:
            last_state = proj_ctx._current_attribute_state_set._states[-1]  # type: ProjectionAttributeState
            in_support_of_trait = self.supporting_attribute.applied_traits.append('is.addedInSupportOf')  # type: CdmTraitReferenceBase
            in_support_of_trait.arguments.append('inSupportOf', last_state._current_resolved_attribute.resolved_name)

        # Create the supporting attribute with the specified 'SupportingAttribute' property as its target and apply the trait 'is.virtual.attribute' to it
        add_trait = ['is.virtual.attribute']
        new_res_attr = self._create_new_resolved_attribute(proj_ctx, attr_ctx_supporting_attr, self.supporting_attribute, added_simple_ref_traits=add_trait)

        # Create a new projection attribute state for the new supporting attribute and add it to the output set
        # There is no previous state for the newly created supporting attribute
        new_PAS = ProjectionAttributeState(proj_output_set._ctx)
        new_PAS._current_resolved_attribute = new_res_attr

        proj_output_set._add(new_PAS)

        return proj_output_set

File: objectModel/Java/objectmodel/src/main/java/com/microsoft/commondatamodel/objectmodel/cdm/projections/CdmOperationAddSupportingAttribute.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

package com.microsoft.commondatamodel.objectmodel.cdm.projections;

import com.microsoft.commondatamodel.objectmodel.cdm.*;
import com.microsoft.commondatamodel.objectmodel.enums.CdmAttributeContextType;
import com.microsoft.commondatamodel.objectmodel.enums.CdmLogCode;
import com.microsoft.commondatamodel.objectmodel.enums.CdmObjectType;
import com.microsoft.commondatamodel.objectmodel.enums.CdmOperationType;
import com.microsoft.commondatamodel.objectmodel.resolvedmodel.ResolvedAttribute;
import com.microsoft.commondatamodel.objectmodel.resolvedmodel.projections.ProjectionAttributeState;
import com.microsoft.commondatamodel.objectmodel.resolvedmodel.projections.ProjectionAttributeStateSet;
import com.microsoft.commondatamodel.objectmodel.resolvedmodel.projections.ProjectionContext;
import com.microsoft.commondatamodel.objectmodel.utilities.*;
import com.microsoft.commondatamodel.objectmodel.utilities.logger.Logger;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Class to handle AddSupportingAttribute operations
 */
public class CdmOperationAddSupportingAttribute extends CdmOperationBase {
    private static final String TAG = CdmOperationAddSupportingAttribute.class.getSimpleName();
    private CdmTypeAttributeDefinition supportingAttribute;

    public CdmOperationAddSupportingAttribute(final CdmCorpusContext ctx) {
        super(ctx);
        this.setObjectType(CdmObjectType.OperationAddSupportingAttributeDef);
        this.setType(CdmOperationType.AddSupportingAttribute);
    }

    @Override
    public CdmObject copy(ResolveOptions resOpt, CdmObject host) {
        if (resOpt == null) {
            resOpt = new ResolveOptions(this, this.getCtx().getCorpus().getDefaultResolutionDirectives());
        }

        CdmOperationAddSupportingAttribute copy = host == null ? new CdmOperationAddSupportingAttribute(this.getCtx()) : (CdmOperationAddSupportingAttribute)host;

        copy.setSupportingAttribute(
                this.getSupportingAttribute() != null
                        ? (CdmTypeAttributeDefinition)this.getSupportingAttribute().copy(resOpt) : null);

        this.copyProj(resOpt, copy);
        return copy;
    }

    public CdmTypeAttributeDefinition getSupportingAttribute() {
        return supportingAttribute;
    }

    public void setSupportingAttribute(final CdmTypeAttributeDefinition supportingAttribute) {
        this.supportingAttribute = supportingAttribute;
    }

    /**
     * @deprecated This function is extremely likely to be removed in the public interface, and not
     * meant to be called externally at all. Please refrain from using it.
     */
    @Override
    @Deprecated
    public Object copyData(final ResolveOptions resOpt, final CopyOptions options) {
        return CdmObjectBase.copyData(this, resOpt, options, CdmOperationAddSupportingAttribute.class);
    }

    @Override
    public String getName() {
        return "operationAddSupportingAttribute";
    }

    /**
     * @deprecated This function is extremely likely to be removed in the public interface, and not
     * meant to be called externally at all. Please refrain from using it.
     */
    @Override
    @Deprecated
    public CdmObjectType getObjectType() {
        return CdmObjectType.OperationAddSupportingAttributeDef;
    }

    @Override
    public boolean validate() {
        ArrayList<String> missingFields = new ArrayList<>();

        if (this.supportingAttribute == null) {
            missingFields.add("supportingAttribute");
        }
        if (missingFields.size() > 0) {
            Logger.error(this.getCtx(), TAG, "validate", this.getAtCorpusPath(), CdmLogCode.ErrValdnIntegrityCheckFailure, this.getAtCorpusPath(), String.join(", ", missingFields.parallelStream().map((s) -> { return String.format("'%s'", s);}).collect(Collectors.toList())));
            return false;
        }
        return true;
    }

    @Override
    public boolean visit(final String pathFrom, final VisitCallback preChildren, final VisitCallback postChildren) {
        String path = this.fetchDeclaredPath(pathFrom);

        if (preChildren != null && preChildren.invoke(this, path)) {
            return false;
        }

        if (postChildren != null && postChildren.invoke(this, path)) {
            return true;
        }

        return false;
    }

    /**
     * @deprecated This function is extremely likely to be removed in the public interface, and not
     * meant to be called externally at all. Please refrain from using it.
     */
    @Override
    @Deprecated
    public ProjectionAttributeStateSet appendProjectionAttributeState(ProjectionContext projCtx, ProjectionAttributeStateSet projOutputSet, CdmAttributeContext attrCtx) {
        // Pass through all the input projection attribute states if there are any
        for (ProjectionAttributeState currentPAS : projCtx.getCurrentAttributeStateSet().getStates()) {
            projOutputSet.add(currentPAS);
        }

        // Create a new attribute context for the operation
        AttributeContextParameters attrCtxOpAddSupportingAttrParam = new AttributeContextParameters();
        attrCtxOpAddSupportingAttrParam.setUnder(attrCtx);
        attrCtxOpAddSupportingAttrParam.setType(CdmAttributeContextType.OperationAddSupportingAttribute);
        attrCtxOpAddSupportingAttrParam.setName("operation/index" + this.getIndex() + "/" + this.getName());
        CdmAttributeContext attrCtxOpAddSupportingAttr = CdmAttributeContext.createChildUnder(projCtx.getProjectionDirective().getResOpt(), attrCtxOpAddSupportingAttrParam);

        // Create a new attribute context for the supporting attribute we will create
        AttributeContextParameters attrCtxSupportingAttrParam = new AttributeContextParameters();
        attrCtxSupportingAttrParam.setUnder(attrCtxOpAddSupportingAttr);
        attrCtxSupportingAttrParam.setType(CdmAttributeContextType.AddedAttributeSupporting);
        attrCtxSupportingAttrParam.setName(this.supportingAttribute.getName());
        CdmAttributeContext attrCtxSupportingAttr = CdmAttributeContext.createChildUnder(projCtx.getProjectionDirective().getResOpt(), attrCtxSupportingAttrParam);
        
        // TODO: this if statement keeps the functionality the same way it works currently in resolution guidance.
        // This should be changed to point to the foreign key attribute instead. 
        // There has to be some design decisions about how this will work and will be done in the next release.
        if (projCtx.getCurrentAttributeStateSet().getStates().size() > 0) {
            int lastIndex = projCtx.getCurrentAttributeStateSet().getStates().size() - 1;
            ProjectionAttributeState lastState = projCtx.getCurrentAttributeStateSet().getStates().get(lastIndex);
            CdmTraitReference inSupportOfTrait = (CdmTraitReference) this.supportingAttribute.getAppliedTraits().add("is.addedInSupportOf");
            inSupportOfTrait.getArguments().add("inSupportOf", lastState.getCurrentResolvedAttribute().getResolvedName());
        }

        // Create the supporting attribute with the specified "supportingAttribute" property as its target and apply the trait "is.virtual.attribute" to it
        List<String> addTrait = new ArrayList<String>(Arrays.asList("is.virtual.attribute"));
        ResolvedAttribute newResAttr = createNewResolvedAttribute(projCtx, attrCtxSupportingAttr, this.supportingAttribute, null, addTrait);

        // Create a new projection attribute state for the new supporting attribute and add it to the output set
        // There is no previous state for the newly created supporting attribute
        ProjectionAttributeState newPAS = new ProjectionAttributeState(projOutputSet.getCtx());
        newPAS.setCurrentResolvedAttribute(newResAttr);

        projOutputSet.add(newPAS);

        return projOutputSet;
    }
}
